src/dist/replication/lib/duplication/duplication_sync_timer.cpp

```diff
@@ -0,0 +1,146 @@
+                     "meta server");
+            return;
+        }
+    }
+
+    auto req = make_unique<duplication_sync_request>();
+    req->node = _stub->primary_address();
+
+    duplication_sync_rpc rpc(std::move(req), RPC_CM_DUPLICATION_SYNC, 3_s);
+    rpc_address meta_server_address(_stub->get_meta_server_address());
+    ddebug_f("duplication_sync to meta({})", meta_server_address.to_string());
+
+    zauto_lock l(_lock);
+    _rpc_task =
+        rpc.call(meta_server_address, &_stub->_tracker, [this, rpc](error_code err) mutable {
+            on_duplication_sync_reply(err, rpc.response());
+        });
+}
+
+void duplication_sync_timer::on_duplication_sync_reply(error_code err,
+                                                       const duplication_sync_response &resp)
+{
+    if (err == ERR_OK && resp.err != ERR_OK) {
+        err = resp.err;
+    }
+    if (err != ERR_OK) {
+        derror_f("on_duplication_sync_reply: err({})", err.to_string());
+    } else {
+        update_duplication_map(resp.dup_map);
+    }
+
+    zauto_lock l(_lock);
+    _rpc_task = nullptr;
+}
+
+void duplication_sync_timer::update_duplication_map(
+    const std::map<int32_t, std::map<int32_t, duplication_entry>> &dup_map)
+{
+    for (replica_ptr &r : get_all_replicas()) {
+        auto it = dup_map.find(r->get_gpid().get_app_id());
+        if (it == dup_map.end()) {
+            // no duplication is assigned to this app
+            r->get_duplication_manager()->update_duplication_map({});
+        } else {
+            r->get_duplication_manager()->update_duplication_map(it->second);
+        }
+    }
+}
+
+duplication_sync_timer::duplication_sync_timer(replica_stub *stub) : _stub(stub) {}
+
+duplication_sync_timer::~duplication_sync_timer() {}
+
+std::vector<replica_ptr> duplication_sync_timer::get_all_primaries()
+{
+    std::vector<replica_ptr> replica_vec;
+    {
+        zauto_read_lock l(_stub->_replicas_lock);
+        for (auto &kv : _stub->_replicas) {
+            replica_ptr r = kv.second;
+            if (r->status() != partition_status::PS_PRIMARY) {
+                continue;
+            }
+            replica_vec.emplace_back(std::move(r));
+        }
+    }
+    return replica_vec;
+}
+
+std::vector<replica_ptr> duplication_sync_timer::get_all_replicas()
+{
+    std::vector<replica_ptr> replica_vec;
+    {
+        zauto_read_lock l(_stub->_replicas_lock);
+        for (auto &kv : _stub->_replicas) {
+            replica_ptr r = kv.second;
+            replica_vec.emplace_back(std::move(r));
+        }
+    }
+    return replica_vec;
+}
+
+void duplication_sync_timer::close()
+{
+    ddebug("stop duplication sync");
+
+    {
+        zauto_lock l(_lock);
+        if (_rpc_task) {
+            _rpc_task->cancel(true);
+            _rpc_task = nullptr;
+        }
+    }
+
+    if (_timer_task) {
+        _timer_task->cancel(true);
+        _timer_task = nullptr;
+    }
+}
+
+void duplication_sync_timer::start()
```

Comment:
When this 'start' in called?

Reply:
https://github.com/XiaoMi/rdsn/pull/281/commits/371947ec89d521f125b2478f208b0ea1ff3b283c

See `replica_stub::initialize_start`, I have committed the code point.